### ISSUE TITLE
'from data_loader_oo import DataContainer' Where is the data_loader_oo?

### ISSUE THREAD

**Issue by 8188** [fteufel/PyTorch-GRU-D#1]
As the title.

**Comment by fteufel** [comment]
Hi,

This was my first project in Pytorch, so it's a bit of a misnomer, it's not what we commonly call a dataloader. Rather, it's a class that serves as container for data saved as `.pkl` after a lenghty preprocessing pipeline. So the only purpose of this thing is to hold numpy arrays as its attributes.

```
class DataContainer():
    """""
    Container class for the input data. Loads the data from a .pkl at filepath, and processes it according to data_sources
    masks large samples before returning the data. TODO: has hardcoded configs right now
    """""

    def __init__(self, filepath, data_sources):
        #Define path to dataset
        datapath = filepath

        #Configs:
        #data_sources = ['labs','vitals','covs']
        num_obs_thres = 10000
        n_mc_smps = 10 
        obs_min = np.max([10, n_mc_smps]) #remove the samples with less than 10 observation values, Futoma's lanczos impl fails when mc_smps > num_obs!
        
        #Load full dataset from pickle
        full_dataset = pickle.load( open( datapath, "rb" ))

        #Unpack dataset:
        variables = full_dataset[0] #list of all used variables
        train_data,validation_data,test_data = full_dataset[1:4]
        #if 'covs' in data_sources:
        train_static_data, validation_static_data, test_static_data = full_dataset[4:7]

        #Masking step (removing outlier patients with too many observations (memory issue) or too few)
        train_data, train_static_data = mask_large_samples(train_data, num_obs_thres, obs_min, static=train_static_data)
        validation_data, validation_static_data = mask_large_samples(validation_data, num_obs_thres, obs_min, static=validation_static_data)
        test_data, test_static_data = mask_large_samples(test_data, num_obs_thres, obs_min, static=test_static_data)

        #data as array for interfacing with evaluation functions
        self.train_data = train_data
        self.test_data = test_data
        self.validation_data = validation_data

        #Checking data split counts
        self.M = len(variables)  
        self.Ntr = len(train_data[0])
        self.Nva = len(validation_data[0])
        self.Ntest = len(test_data[0])
        self.n_covs = 0

        
        #Unpacking compact format of data splits (Train and Val for now)
        self.values_tr = train_data[0]; self.values_va = validation_data[0]; self.values_test = test_data[0]
        self.times_tr = train_data[1]; self.times_va = validation_data[1]; self.times_test = test_data[1]
        self.ind_lvs_tr = train_data[2]; self.ind_lvs_va = validation_data[2]; self.ind_lvs_test = test_data[2]
        self.ind_times_tr = train_data[3]; self.ind_times_va = validation_data[3]; self.ind_times_test = test_data[3]
        self.labels_tr = train_data[4]; self.labels_va = validation_data[4]; self.labels_test = test_data[4]
        self.num_rnn_grid_times_tr = train_data[5]; self.num_rnn_grid_times_va = validation_data[5]; self.num_rnn_grid_times_test = test_data[5]
        self.rnn_grid_times_tr = train_data[6]; self.rnn_grid_times_va = validation_data[6]; self.rnn_grid_times_test = test_data[6]
        self.num_obs_times_tr = train_data[7]; self.num_obs_times_va = validation_data[7]; self.num_obs_times_test = test_data[7]
        self.num_obs_values_tr = train_data[8]; self.num_obs_values_va = validation_data[8]; self.num_obs_values_test = test_data[8]
        self.onset_times_tr = train_data[9]; self.onset_times_va = validation_data[9]; self.onset_times_test = test_data[9]
        if 'covs' in data_sources:
            self.covs_tr = train_static_data
            self.n_covs = train_static_data.shape[1]
            self.covs_va = validation_static_data #; covs_te = test_static_data
            self.covs_test = test_static_data    

```

**Comment by simonbing** [comment]
Could you elaborate what is stored in `num_obs_times_tr` and what the expected dimensions are? I am trying to train on my own data and it is not quite clear what is expected here.

**Comment by fteufel** [comment]
Sure!
The data I trained on was obtained from the preprocessing in https://github.com/BorgwardtLab/mgp-tcn.
The pipeline there in `main_preprocessing_mgp_tcn.py` produces the pickle files that `DataContainer` consumes. Can't just share the file with you unfortunately, there's restrictions on the data.

Anyway, I opened the pickle file I still had. It contains a list with the following types:
```
[numpy.ndarray, list, list, list, numpy.ndarray, numpy.ndarray, numpy.ndarray]
```
This contains the following:
```
variable_names, train_data, validation_data, test_data, train_static_data, validation_static_data, test_static_data
```

`train_data`, `validation_data` and `test_data` each contain 10 arrays:
```
[x.shape for x in train_data]
[(4968,),
 (4968,),
 (4968,),
 (4968,),
 (4968,),
 (4968,),
 (4968,),
 (4968,),
 (4968,),
 (4968,)]
```

This is where it gets messy. As the time series are not evenly spaced and have different numbers of observations, those are jagged arrays. Each entry contains the values for one patient.
```
self.values_tr = train_data[0] #jagged. the actual measured values, float, shape (n_measurements)
self.times_tr = train_data[1]  #jagged. the actual time points, float. shape (n_timepoints)
self.ind_lvs_tr = train_data[2] #jagged. the channel of each measured value, int. shape (n_measurements)
self.ind_times_tr = train_data[3] #jagged. the time point of each measured value. int. shape(n_measurements). This gives the index into times_tr[x] to retrieve the timepoint for each value in values_tr[x]
self.labels_tr = train_data[4] #the patient label. int (0,1)
self.num_rnn_grid_times_tr = train_data[5] #the maximum grid time for mgp-tcn. don't think this was used here. int
```

Example: patient 0 has 296 measured values in `values_tr[0]`. These were taken at 49 time points in `times_tr[0]`. `ind_lvs_tr[0]` contains 296 ints indicating to which channel each measured value belongs. `ind_times_tr[0]` contains 296 ints, indexing those into `times_tr[0]` returns the time point of each measurement.

I know this probably all sounds unnecessarily confusing, but it really does keep the file size small. Hope it helps.

**Comment by fteufel** [comment]
It probably makes more sense to ignore all of the above if you have your own data and just go straight to converting your own format to the input format. 

The model takes a 4D tensor as input with shape `(batch_size, 3, n_features, seq_length)`.
Along dim 1, `0` is the data matrix X, `1` is the masking matrix and `2` is the delta T matrix. The data matrix has 0 where there was no value observed, masking and delta T are as given in Figure 2 of the paper.
